hadoop-hdds/docs/content/design/container-reconciliation.md

```diff
@@ -0,0 +1,344 @@
+
+There may be cases where a container is critically under-replicated and we need a good copy of the data as soon as possible. Meanwhile, reconciliation could remain blocked for an unrelated reason. We can provide a way to pause/unpause reconciliation between chunk writes to do reasonably consistent container exports.
+
+#### On block delete
+
+- When SCM sends block delete commands to datanodes, update the merkle tree when the datanode block deleting service runs and processes that block.
+
+    - Merkle tree update should be done before RocksDB update to make sure it is persisted in case of a failure during delete.
+
+    - Merkle tree update should not be done when datanodes first receives the block delete command from SCM, because this command only adds the delete block proto to the container. It does not iterate the blocks to be deleted so we should not add that additional step here.
+
+#### On Container delete for container being reconciled
+
+- Provide a way to interrupt reconciliation and allow container delete to proceed. 
+
+## Backwards Compatibility
+
+Since the scanner can generate the container merkle trees in the background, existing containers created before this feature will still be eligible for reconciliation. These old containers may not have all of their block deletes present in the merkle tree, however, which could cause some false positives about missing blocks on upgrade if one node had already deleted blocks from a container before the upgrade, and another datanode has not yet processed the delete of those blocks.
+
+This can be mitigated by:
+1. Having datanodes delete blocks from their container replica on reconciliation that another replica has marked as deleted.
+2. Having SCM periodically ask datanodes to reconcile otherwise matching containers when there are no other mismatched containers in the cluster.
+
+The sequence of events would look like this. Assume software v1 does not have container reconciliation, but v2 does.
+1. Datanode 1 in v1 deletes block 1 in container 1. This does not leave a tombstone entry because v1 does not have container reconciliation.
+2. Cluster is upgraded to v2.
+3. Reconciliation is triggered between datanodes 1 and 2 for container 1. Datanode 2 has not yet deleted block 1.
+4. Datanode 1 will add block 1 to its container when reconciling with datanode 2.
+    - If we stop here, then block 1 will remain orphaned on datanode 1.
+5. Datanode 2 deletes block 1 and adds a tombstone entry for it since the cluster is running software v2 which has container reconciliation.
+6. Since container checksums disregard deleted blocks, container 1 will be seen as matching from SCM's point of view. However, periodic reconciliation requests for closed containers will still eventually ask these two replicas to reconcile.
+7. Datanode 1 learns that block 1 was deleted from datanode 2, so it moves the block metadata to the deleted table in RocksDB
```

Comment:
There is a `bool deleted` flag in the merkle tree proto that will be updated for each block delete as a tombstone entry. When datanodes use the proposed API to pull the Merkle tree, they can see that their peer deleted the block so it is not missing.

Reply:
Understood. Datanode 1 needs to wait for SCM to send the "reconcileContainer" API call to it. Once this call is received, Datanode 1 will retrieve the Merkle tree from Datanode 2. This allows Datanode 1 to know that block 1 has been deleted, right?
However, this process may be time-consuming, as SCM might not sending "reconcileContainer" solely due to an orphan block.

Reply:
@xichen01 this is good point, I think similar to a node level scan we will need a SCM level scan where every container in the system is reconciled once in `x` number of days even if the checksums match. The other option is what @szetszwo recommended is to have a separate delete hash in the Merkle Tree to detect missing deletions.

Reply:
Yes we should not depend on reconcile purely for deleting blocks. Main block delete flow will still be active. There are corner cases where the current delete flow can leave orphan blocks, so it would be nice to have for the reconciliation process to handle this, although it is not a primary goal.